### ISSUE TITLE
[Bug] NaNs in CUDA Graph padded regions corrupt (unpadded) scales in nvfp4 quant functions

### ISSUE THREAD

**Issue by tlrmchlsmth** [flashinfer-ai/flashinfer#3057]
#### The bug:
FlashInfer's `silu_and_mul_scaled_nvfp4_experts_quantize` and `scaled_fp4_grouped_quantize` kernels compute per-row FP4 quantization scales, but in certain configurations (sparse token counts — e.g., 8 real tokens across 8 experts with 64 max slots), the scale computation reads data from padding rows beyond masked_m. When those padding rows contain NaN or garbage from stale CUDA graph workspace buffers, the scale values for real token rows get corrupted.

These sparse configurations happen in EP+DP serving due to DP padding to the max toks DP ranks

The corruption produces wrong finite values (not NaN) — zeroed or incorrect scales that silently degrade quantization accuracy.

These kernels are used in the `FlashInferCuteDSLBatchedExperts` MoE path: `scaled_fp4_grouped_quantize` quantizes activations before up_gate, and `silu_and_mul_scaled_nvfp4_experts_quantize` does silu_mul_quant between up_gate and down.

#### See also:
https://github.com/vllm-project/vllm/pull/39743

#### Untested vibe-coded (one shot) fix:
https://github.com/flashinfer-ai/flashinfer/compare/main...tlrmchlsmth:flashinfer:fix/nvfp4-expert-quant-mask-warp-sync?expand=1

**Comment by leonardHONG** [comment]
Hi @tlrmchlsmth, I saw your branch for this and I have a Pro 6000 . I'd be happy to help put together a NaN-padding repro, verify it on hardware, and send a PR if that's useful, unless you're already planning to push a fix yourself.

@kahyunnam I also noticed this was recently self-assigned. If you're already on it, I'm happy to defer.

**Comment by kahyunnam** [comment]
@leonardHONG I have not gotten around to it; please feel welcome to take a stab at this 👍

**Comment by leonardHONG** [comment]
Picked this up per @kahyunnam's handoff. Cherry-picked Tyler's patch onto a local verify branch and ran another round of triage on both `upstream` and `fix-3057`. Probes split into:

- `dirty_eager` — eager call with NaN-padded input
- `dirty_graph` — CUDA graph capture, NaN-padded static input
- `polluted_graph` — graph capture + replay after NaN-poisoning the static input

Low-level checks reran with zero-initialized output buffers.

On RTX PRO 6000 I still don't see an upstream/fix split:

- Corrected real-row oracle: 0 (FP4 byte-equal + unswizzled SF byte-equal on real rows)
- Low-level physical outputs: 0 diff
- The diff I do see is only at the public layer, starts at `dirty_graph` (not `dirty_eager`), is confined to padding rows, and disappears when the public-like outputs are zero-initialized.

So my read is that the stable signal here is public-wrapper exposure of uninitialized / don't-care padding cells in graph-captured static outputs, not the real-row corruption signal I would expect this patch to split on locally. I can localize the signal to the wrapper layer, but I don't have local evidence for an upstream-fail / fix-pass split on the kernel real-row bug.

The patch itself looks reasonable from a code-review pass (partial-warp early-exit + full-mask `__shfl_xor_sync` UB → masked lanes participate with zeroed input). I just couldn't get a before/after numerical diff on this hardware.

Happy to share probe scripts + logs (`E=8`, `M=64`, `masked_m=[1]*8`, bf16, hidden=1024, multi-replay + NaN allocator poisoning, `prime_silu` sequencing) if useful. If you have a vLLM-side repro pointing at `FlashInferCuteDSLBatchedExperts` / DeepEP LL, I can run it on Pro 6000.

**Comment by leonardHONG** [comment]
Quick cross-link for context: this maps to item (2) in `vllm-project/vllm#40047`; my current probe is still an isolated-kernel surface and doesn’t exercise item (5) there — the upstream CUDA-graph NaN source — which may explain why I’m not getting a local real-row split here.

**Comment by tlrmchlsmth** [comment]
Two things that might affect the repro:
* I'm running on GB200 so a different compute capability than Pro 6000
* I only saw the corruption with larger amounts of padding (>= 16 tokens of padding NaNs), which only happens in the DP case

**Comment by kahyunnam** [comment]
Also having a hard time repro-ing the real row corruption with main branch on B200, as a standalone bug for the quantization op. What I can repro is similar to @leonardHONG 's finding that the real row is intact, while padding rows are corrupted.

I see that this is a 2 bug chain: 
1. #3057 (quant) — leaves NaN in the output scale padding (what we can fix by zero-ing). On its own it doesn't corrupt real rows.
2. #3103 (GEMM) — consumes those scales and leaks any NaN scale across all experts' real outputs.

I was able to repro 2 as described.

**Comment by kahyunnam** [comment]
Not sure this is an actual real-row-corruption bug, since we couldn't reproduce real-row corruption here (only stale NaN in unwritten padding rows), so the zero-fill solution in #3571 is more of a defensive hygiene rather a correctness fix. The real MoE corruption appears to be #3103 (grouped_gemm_nt_masked output-layout issue), resolved in #3574.

**Comment by tlrmchlsmth** [comment]
@kahyunnam can you try this test specifically? https://github.com/tlrmchlsmth/vllm/pull/33/changes#r3075733024

**Comment by tlrmchlsmth** [comment]
Note that it only repros when there are CUDA Graphs being used with a large amount of padding. This is a case that slips between the cracks during normal testing but can happen in large scale serving due to DP load imbalance

**Comment by kahyunnam** [comment]
Thanks @tlrmchlsmth, I ported that test (`test_flashinfer_silu_quant_cross_row_corruption`) into a standalone script (same logic and comparison, just w/o the vLLM deps) and ran it on B200 / CUDA 13.0 / latest FlashInfer main. On the unfixed quant kernel it does report `FAILED` for the larger-padding configs (`E16_m64`, `E32_m128`, `E8_m64_sparse`, `E4_m16_1tok`). But the failure seems to be from the test comparing the wrong axis, not real-token corruption, and the #3571 zero-fill makes the test pass.

<details>
<summary>repro_tlrm_silu_test.py — ported standalone test (no vLLM deps)</summary>

```python
import torch
from flashinfer import silu_and_mul_scaled_nvfp4_experts_quantize
CONFIGS = [
    (4, 8, 4, "E4_m8"),
    (8, 16, 8, "E8_m16"),
    (8, 32, 16, "E8_m32"),
    (16, 64, 32, "E16_m64"),
    (32, 128, 64, "E32_m128"),
    (8, 64, 8, "E8_m64_sparse"),
    (4, 16, 1, "E4_m16_1tok"),
]
def run_one(num_experts, m, num_real):
    device = "cuda"
    n = 512
    base = num_real // num_experts
    remainder = num_real % num_experts
    masked_m = torch.tensor(
        [base + (1 if i < remainder else 0) for i in range(num_experts)],
        dtype=torch.int32, device=device,
    )
    a2_global_scale = torch.ones(num_experts, dtype=torch.float32, device=device)
    workspace = torch.randn(num_experts, m, 2 * n, dtype=torch.bfloat16, device=device) * 0.1
    for e in range(num_experts):
        real = masked_m[e].item()
        if real < m:
            workspace[e, real:] = float("nan")
    workspace_clean = workspace.clone()
    for e in range(num_experts):
        real = masked_m[e].item()
        if real < m:
            workspace_clean[e, real:] = 0.0
    diq_clean, sf_clean = silu_and_mul_scaled_nvfp4_experts_quantize(
        workspace_clean, masked_m, a2_global_scale
    )
    silu_and_mul_scaled_nvfp4_experts_quantize(workspace.clone(), masked_m, a2_global_scale)
    torch.cuda.synchronize()
    graph = torch.cuda.CUDAGraph()
    workspace_input = workspace.clone()
    with torch.cuda.graph(graph):
        diq, sf = silu_and_mul_scaled_nvfp4_experts_quantize(
            workspace_input, masked_m, a2_global_scale
        )
    workspace_input.fill_(float("nan")); graph.replay(); torch.cuda.synchronize()
    workspace_input.copy_(workspace); graph.replay(); torch.cuda.synchronize()
    corrupted = False
    for e in range(num_experts):
        real = masked_m[e].item()
        if real == 0:
            continue
        if not torch.equal(diq_clean[e, :real], diq[e, :real]):
            corrupted = True; break
        if not torch.equal(sf_clean[e, :real], sf[e, :real]):
            corrupted = True; break
    return corrupted
def main():
    cap = torch.cuda.get_device_capability("cuda:0")
    print(f"torch={torch.__version__} cuda={torch.version.cuda} "
          f"GPU={torch.cuda.get_device_name('cuda:0')} cc={cap[0]}.{cap[1]}\n")
    any_fail = False
    for (E, m, real, name) in CONFIGS:
        corrupted = run_one(E, m, real)
        any_fail |= corrupted
        print(f"  {name:14s} (E={E},m={m},real={real}) -> "
              f"{'FAILED' if corrupted else 'PASSED'}")
    print("\n" + ("SOME CONFIGS FAILED" if any_fail else "ALL PASSED"))
if __name__ == "__main__":
    main()
```
</details>

<details>
<summary>repro_tlrm_dual_oracle.py — same setup, raw vs correctly-indexed real-token check</summary>

```python
import torch
from flashinfer import silu_and_mul_scaled_nvfp4_experts_quantize
CONFIGS = [
    (8, 64, 8, "E8_m64_sparse"),
    (16, 64, 32, "E16_m64"),
    (32, 128, 64, "E32_m128"),
    (4, 16, 1, "E4_m16_1tok"),
]
def unswizzle_full(sf_tile, padded_m, feat, vec=16):
    factor = vec * 4
    num_m = (padded_m + 127) // 128
    num_k = (feat + factor - 1) // factor
    r = sf_tile.view(num_m, num_k, 32, 4, 4).transpose(1, 3)
    r = r.reshape(num_m * 32 * 4, num_k * 4)
    return r[:padded_m, : (feat // vec)].contiguous()
def diff_rows(a, b):
    af = torch.nan_to_num(a.float(), nan=-1.0)
    bf = torch.nan_to_num(b.float(), nan=-2.0)
    rd = (af != bf).reshape(a.shape[0], -1).any(dim=1)
    return torch.nonzero(rd).flatten().tolist()
def run_one(num_experts, m, num_real):
    device = "cuda"
    n = 512
    pm = ((m + 127) // 128) * 128
    base = num_real // num_experts
    rem = num_real % num_experts
    masked_m = torch.tensor(
        [base + (1 if i < rem else 0) for i in range(num_experts)],
        dtype=torch.int32, device=device,
    )
    gs = torch.ones(num_experts, dtype=torch.float32, device=device)
    workspace = torch.randn(num_experts, m, 2 * n, dtype=torch.bfloat16, device=device) * 0.1
    for e in range(num_experts):
        r = masked_m[e].item()
        if r < m:
            workspace[e, r:] = float("nan")
    clean = workspace.clone()
    for e in range(num_experts):
        r = masked_m[e].item()
        if r < m:
            clean[e, r:] = 0.0
    diq_clean, sf_clean = silu_and_mul_scaled_nvfp4_experts_quantize(clean, masked_m, gs)
    silu_and_mul_scaled_nvfp4_experts_quantize(workspace.clone(), masked_m, gs)
    torch.cuda.synchronize()
    g = torch.cuda.CUDAGraph()
    ws = workspace.clone()
    with torch.cuda.graph(g):
        diq, sf = silu_and_mul_scaled_nvfp4_experts_quantize(ws, masked_m, gs)
    ws.fill_(float("nan")); g.replay(); torch.cuda.synchronize()
    ws.copy_(workspace); g.replay(); torch.cuda.synchronize()
    # Raw slice (the test's check)
    raw_fail = False
    for e in range(num_experts):
        r = masked_m[e].item()
        if r == 0:
            continue
        if not torch.equal(diq_clean[e, :r], diq[e, :r]) or not torch.equal(
            sf_clean[e, :r], sf[e, :r]
        ):
            raw_fail = True; break
    # Correctly-indexed real tokens
    diq_p = diq.permute(2, 0, 1); diqc_p = diq_clean.permute(2, 0, 1)
    sf_p = sf.permute(5, 2, 4, 0, 1, 3); sfc_p = sf_clean.permute(5, 2, 4, 0, 1, 3)
    proper_fail = False
    min_packed = min_scale = 10**9
    for e in range(num_experts):
        r = masked_m[e].item()
        if r == 0:
            continue
        pr = diff_rows(diqc_p[e, :r], diq_p[e, :r])
        sc = diff_rows(unswizzle_full(sfc_p[e], pm, n)[:r], unswizzle_full(sf_p[e], pm, n)[:r])
        if pr: min_packed = min(min_packed, min(pr))
        if sc: min_scale = min(min_scale, min(sc))
        if pr or sc: proper_fail = True
    return raw_fail, proper_fail, min_packed, min_scale
def main():
    cap = torch.cuda.get_device_capability("cuda:0")
    print(f"GPU={torch.cuda.get_device_name('cuda:0')} cc={cap[0]}.{cap[1]}\n")
    for (E, m, real, name) in CONFIGS:
        raw_fail, proper_fail, mp, ms = run_one(E, m, real)
        mp_s = "none" if mp == 10**9 else str(mp)
        ms_s = "none" if ms == 10**9 else str(ms)
        verdict = ("REAL-TOKEN CORRUPTION" if proper_fail
                   else ("padding/swizzle artifact only" if raw_fail else "fully clean"))
        print(f"  {name:14s}: RAW={'FAIL' if raw_fail else 'ok'}  "
              f"REAL={'FAIL' if proper_fail else 'ok'}  "
              f"[real-row diff: packed={mp_s} scale={ms_s}]  -> {verdict}")
if __name__ == "__main__":
    main()
```
</details>

### Why the comparison mis-fires on this build

`silu_and_mul_scaled_nvfp4_experts_quantize` returns with the expert (`l`) dim **last**:
- `diq` → shape `[m, k/2, l]`
- `sf`  → shape `[32, 4, padded_m//128, 4, padded_k//4, l]`

(end of the sm100 path in `flashinfer/quantization/fp4_quantization.py`):

```python
output = output.permute(1, 2, 0)                       # -> [m, k/2, l]
output_scales = output_scales.view(torch.float8_e4m3fn).view(
    l, padded_m // 128, padded_k // 4, 32, 4, 4)
output_scales = output_scales.permute(3, 4, 1, 5, 2, 0) # -> [32, 4, .., l]
return output, output_scales
```

The test compares (lines 1916 & 1919):

```python
if not torch.equal(diq_clean[e, :real], diq[e, :real]): ...
if not torch.equal(sf_clean[e, :real], sf[e, :real]):   ...
```

i.e. it treats **axis 0 as the expert index** `e`. But on this build axis 0 of `diq` is `m`, and axis 0 of `sf` is the swizzle-`32`. So `diq[e, :real]` / `sf[e, :real]` select token-slots / swizzle components — **not** expert `e`'s real tokens — and those slices include unwritten swizzle/padding cells that still hold the stale NaN from the pollution replay. So it flags stale padding, not corrupted real data.

### Proof on B200 (w/ your exact pollution setup, `randn*0.1` workspace, NaN padding, CUDA-graph all-NaN replay → real replay)

For each config I checked both the raw `[e,:real]` slice (what the test does) and the **correctly-indexed real tokens** (permute `l` to the front, slice `[:masked_m]`; unswizzle the SF to logical rows):

| config | raw `[e,:real]` (test's check) | correctly-indexed real tokens |
|---|---|---|
| E8_m64_sparse | FAIL | **ok** (packed diff = none, scale diff = none) |
| E16_m64 | FAIL | **ok** |
| E32_m128 | FAIL | **ok** |
| E4_m16_1tok | FAIL | **ok** |

Real tokens are bit-identical between the NaN-padded and clean runs for every config (consistent with @leonardHONG's RTX PRO 6000 triage).

### #3571 makes the test pass

Since the test diffs the polluted run against a clean run, zero-filling the output padding makes those cells deterministic in both. Re-running the same test with #3571 applied on B200:

```
E4_m8 PASSED · E8_m16 PASSED · E8_m32 PASSED · E16_m64 PASSED
E32_m128 PASSED · E8_m64_sparse PASSED · E4_m16_1tok PASSED
ALL PASSED
```

### One caveat / ask
This `[m, k/2, l]` (expert-`l`-**last**) return layout has been in place since #2443 (2026-03-18) and unchanged — it predates this test by a few weeks — so the same axis-0 mis-indexing very likely applied on your GB200 run as well, which would make that failure the same padding artifact rather than real-token corruption. To be certain on GB200, could you re-run with the correctly-indexed check (permute `l` to the front, slice `[:masked_m]`, and unswizzle the SF to logical rows — script below) and confirm whether any *real-token* values actually differ? On B200 / Pro-6000 they don't, and #3571 makes the padding deterministic either way.

**Comment by kahyunnam** [comment]
Closing this! @tlrmchlsmth please feel welcome to reopen or open a new issue if this problem persists

**Comment by tlrmchlsmth** [comment]
Thank you for digging in and resolving these!
